src/FastSerialization/StreamReaderWriter.cs

```diff
@@ -656,6 +656,41 @@ public override void Goto(StreamLabel label)
         /// </summary>
         public override long Length { get { return inputStream.Length; } }
         public override bool HasLength { get { return inputStream.CanSeek; } }
+
+        public override void Read(byte[] data, int offset, int length)
+        {
+            // The base class is constrained to only handle reads that are <= size of the cache
+            // For larger reads we can take what is available in the cache and satisfy the
+            // remainder from the stream
+            if (length > bytes.Length)
+            {
+                int cacheBytes = endPosition - position;
+                Buffer.BlockCopy(bytes, position, data, offset, cacheBytes);
+                int bytesRead = cacheBytes;
+                while (bytesRead < length)
+                {
+                    int count = inputStream.Read(data, offset + bytesRead, length - bytesRead);
+                    inputStreamBytesRead += (uint)count;
+                    if (count == 0)
```

Comment:
I think it's possible here to get into a state where the stream doesn't return the requested number of bytes but does return more than 0.  In this case, wouldn't inputStream.Read return a number > 0 but < count?  If so, then I think you would want to update the Goto line to be Goto(Current.Add(bytesRead));

Reply:
I assume where you said 'count' you meant (length - bytesRead) which is the number of bytes we request to read from the stream.
If it returns a value > 0 and less than (length - bytesRead) then we execute: bytesRead += count, loop back to while(bytesRead < length) and start another iteration of the loop because the condition is still true. We should only exit the while loop if either:

a) we throw the exception that we read past end of stream
b) bytesRead >= length (and bytesRead should never go bigger than length unless the stream has a bug and returned more bytes than we requested)


When we execute the Goto, bytesRead == length, unless I am still missing something?

Reply:
Yes, you're right.  Not sure what I missed there.